pkg/composer/up_service.go

```diff
@@ -0,0 +1,141 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package composer
+
+import (
+	"context"
+	"os"
+	"os/exec"
+	"os/signal"
+	"strings"
+
+	"github.com/containerd/nerdctl/pkg/composer/pipetagger"
+	"github.com/containerd/nerdctl/pkg/composer/serviceparser"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+func (c *Composer) upServices(ctx context.Context, parsedServices []*serviceparser.Service, uo UpOptions) error {
+	if len(parsedServices) == 0 {
+		return errors.New("no service was provided")
+	}
+
+	// TODO: parallelize loop for ensuring images (make sure not to mess up tty)
+	for _, ps := range parsedServices {
+		if err := c.ensureServiceImage(ctx, ps); err != nil {
+			return err
+		}
+	}
+
+	var logTagMaxLen int
+	type containerState struct {
+		name   string
+		logTag string
+		logCmd *exec.Cmd
+	}
+
+	containerStates := make(map[string]containerState, len(parsedServices)) // key: containerID
+	for _, ps := range parsedServices {
+		for _, container := range ps.Containers {
+			containerID, err := c.upServiceContainer(ctx, container)
+			if err != nil {
+				return err
+			}
+			logTag := strings.TrimPrefix(container.Name, c.project.Name+"_")
+			if l := len(logTag); l > logTagMaxLen {
+				logTagMaxLen = l
+			}
+			containerStates[containerID] = containerState{
+				name:   container.Name,
+				logTag: logTag,
+			}
+		}
+	}
+
+	if uo.Detach {
+		return nil
+	}
+
+	logrus.Info("Attaching to logs")
+	for id, state := range containerStates {
```

Comment:
I try it in my local.

One terminal A is `_output/nerdctl compose -f ./examples/compose-wordpress/docker-compose.yaml up` and wait for seconds. Open other terminal B is for `_output/nerdctl compose -f ./examples/compose-wordpress/docker-compose.yaml down` and the terminal A still attach logs.

Reply:
maybe we can use `wait` action for watching the container. it can be follow-up.

Reply:
fixed. We don't have `nerdctl wait <CONTAINER>` yet, so wait for exit of `nerdctl logs -f <CONTAINER>`.